cmd/applyconfig/applyconfig.go

```diff
@@ -298,14 +342,68 @@ func init() {
 	logrus.SetFormatter(logrusutil.NewCensoringFormatter(logrus.StandardLogger().Formatter, secrets.getSecrets))
 }
 
+type ocVersionOutput struct {
+	Openshift string `json:"openshiftVersion"`
+}
+
+func selectDryRun(v ocVersionOutput) dryRunMethod {
+	minOpenshift := version.Must(version.NewVersion("4.5.0"))
+
+	// `oc` against `api.ci` does not contain the openshift version field, so if it's missing, we need to assume
+	// we're running against something old
+	if v.Openshift == "" {
+		logrus.Warning("oc version did not provide openshift version: guessing too old, will continue with client dry-run")
+		return dryClient
+	}
+
+	// Generally we want to prefer the stronger (server) validation unless
+	// we *know* we need to use the weak, client one -> so use server on errors
+	openshift, err := version.NewVersion(v.Openshift)
+	if err != nil {
+		logrus.WithError(err).Warning("Cannot parse openshift version; will continue with server dry-run")
+		return dryServer
+	}
+
+	if openshift.GreaterThanOrEqual(minOpenshift) {
+		return dryServer
+	}
+
+	return dryClient
+}
+
 func main() {
 	o := gatherOptions()
-	var hadErr bool
 
 	if !o.user.beenSet {
 		o.user.val = defaultAdminUser
 	}
 
+	if o.dryRun == dryAuto {
+		logrus.Info("Detecting dry run capabilities")
+		executor := commandExecutor{}
+		// Generally we want to always prefer stronger (server) validation unless
+		// we *know* we need to use the weak, client one -> so use server on errors
+		cmd := makeOcCommand(ocVersion, o.kubeConfig, o.context, "", o.user.val, "--output=json")
+		out, err := executor.runAndCheck(cmd, "detect server version")
+		if err == nil {
```

Comment:
This is really hard to read, requires two levels of conditional logic and two places where we do defaulting. Please just factor all the detection logic into a new func that returns the dry run method or an error and make the caller log that error and default to server if it occurs

Reply:
Addressed. I refactored it slightly differently. Extracting to a func allows using guard clause which reads quite nicely, so I did that.